strawberryfields/tdm/tdmprogram.py

```diff
@@ -0,0 +1,327 @@
+    """Returns the registers from q required by a command"""
+    reg_getter = itemgetter(*[r.ind for r in cmd.reg])
+    modes = reg_getter(q)
+
+    if not isinstance(modes, tuple):
+        modes = (modes,)
+
+    return modes
+
+
+def validate_measurements(circuit, N):
+    """Validate the TDM program measurements are correct"""
+    spatial_modes = 0
+    measurement_reg = []
+
+    for cmd in circuit:
+        if isinstance(cmd.op, ops.Measurement):
+            modes = [r.ind for r in cmd.reg]
+            measurement_reg.extend(modes)
+            spatial_modes += len(modes)
+
+    if not spatial_modes:
+        raise ValueError("Must be at least one measurement.")
+
+    if spatial_modes is not len(N):
+        raise ValueError("Number of measurement operators must match number of spatial modes.")
+
+    return spatial_modes
+
+
+def input_check(args, copies):
+    """Checks the input arguments have consistent dimensions"""
+    # check if all lists are of equal length
+    param_lengths = [len(param) for param in args]
+    if len(set(param_lengths)) != 1:
+        raise ValueError("Gate-parameter lists must be of equal length.")
+
+    # check copies
+    if not isinstance(copies, int) or copies < 1:
+        raise TypeError("Number of copies must be a positive integer.")
+
+
+def _get_mode_order(num_of_values, N):
+    """Get the order by which the modes where measured"""
+    N = N.copy()
+
+    all_modes = []
+    for i in range(len(N)):
+        ra = list(range(sum(N[:i]), sum(N[: i + 1])))
+        all_modes.append(ra * ceil(max(N) / len(ra)))
+
+    mode_order = [i for j in zip(*all_modes) for i in j]
+    mode_order *= ceil(num_of_values / len(mode_order))
+
+    return mode_order[:num_of_values]
+
+
+def reshape_samples(all_samples, modes, N):
+    """Reshapes the samples dict so that they have the expected correct shape
+
+    Reshapes the samples dict so that they the measured modes are the ones
+    defined to be measured in the circuit instead of being spread over a larger
+    number of modes due to the mode-shifting occuring in TDMProgram.
+
+    Args:
+        all_samples (dict[int, list]): the raw measured samples
+        modes (Sequence[int]): the modes that are measured in the circuit
+        N (Sequence[int]): the number of used modes per belt/spatial modes
+
+    Returns:
+        new_samples (dict[int, list]): the re-shaped samples
+    """
+    num_of_values = len([i for j in all_samples.values() for i in j])
+    mode_order = _get_mode_order(num_of_values, N)
+
+    # go backwards through all_samples and add them into the correct mode
+    new_samples = dict()
+    for i, m in enumerate(mode_order):
+        idx = modes[i % len(N)]
+        if idx not in new_samples:
+            new_samples[idx] = []
+        new_samples[idx].append(all_samples[m].pop(0))
+    return new_samples
+
+
+class TDMProgram(sf.Program):
+    r"""Represents a photonic quantum circuit in the time domain encoding.
```

Comment:
@fab1an-q I had a go beginning to fill out the TDMProgram docstring, could you take a look and add/edit it as needed? There are a few parts I didn't know fully how to describe :)

Reply:
The build docs are visible here: https://strawberryfields--440.org.readthedocs.build/en/440/code/api/strawberryfields.tdm.TDMProgram.html

They will be automatically re-built after every commit

Reply:
One thing that we haven't explained yet is how to assign the proper qumode indices to the individual concurrent modes: 

- For each spatial mode the number of concurrent modes should equal: 1 + number of time bins in a potential delay line in this spatial mode.
- Divide the total number of concurrent modes into the individual spatial modes with the indices filling up spatial mode A before going to spatial mode B and so on. For instance, if a setup has
2 concurrent modes in spatial mode A,
1 concurrent mode in spatial mode B,
3 concurrent modes in spatial mode C,
then
`q[0]`, `q[1]` will be assigned to A,
`q[2]` will be assigned to B,
`q[3]`, `q[4]`, `q[5]` will be assigned to C
- In each spatial mode, the lowest index corresponds to the oldest mode (the one closest to the detector), whereas the highest index corresponds to the youngest mode (the one being produced at the light source).

Not sure if these instructions should be part of the docstring or rather occur somewhere else? Also, I think a simple example drawing could be really helpful for the user.

Reply:
Hi @fab1an-q . Not sure if this should go in the docstring but it is super useful for the tutorial that should come along with this.

Reply:
Users could benefit from some of our examples in `test_tdmprogram.py` as well. And we also have examples of how we can plot quadrature correlations and so on. This is obviously not relevant for the docstring, but maybe there is another place these examples could fit into?

Reply:
Absolutely! They are the basis for the tutorial.  I will continue working on it this week and once it is in reasonable shape will share it with everyone.

Reply:
Oh ok. I kind of thought the docstring already were the tutorial. Thanks @nquesada!

Reply:
> One thing that we haven't explained yet is how to assign the proper qumode indices to the individual concurrent modes:

It depends how 'core' this is to understand. What we can always do is add more information to the bottom of the docstring, in a

```rest
.. UsageDetails::
```
indented section. It will look like this: https://pennylane.readthedocs.io/en/stable/code/api/pennylane.templates.layers.BasicEntanglerLayers.html (and be minimized by default).

> This is obviously not relevant for the docstring, but maybe there is another place these examples could fit into?

Could these go in the tutorials?

Reply:
@fab1an-q think of the docstring as a quick reference page; e.g., condensed reference information showing everything you need to know about the function.

The tutorial is a separate page, that is more 'telling a story'. That is, showing off the functionality via a set of fully worked through examples (including any theory!)

Reply:
Thanks @josh146! I admit I have misinterpreted the docstring as tutorial. I now understand that it's not the whole story. Yes, all my suggestions above seem to fit well into the tutorial then.